common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -270,6 +279,123 @@ public byte[] getBytes() {
+    return switch (firstByte) {
+      case (byte) 0xE0 -> (byte) 0xA0 <= b && b <= (byte) 0xBF;
+      case (byte) 0xED -> (byte) 0x80 <= b && b <= (byte) 0x9F;
+      case (byte) 0xF0 -> (byte) 0x90 <= b && b <= (byte) 0xBF;
+      case (byte) 0xF4 -> (byte) 0x80 <= b && b <= (byte) 0x8F;
+      default -> isValidContinuationByte(b);
+    };
+  }
+
+  private static final byte[] UNICODE_REPLACEMENT_CHARACTER =
+    new byte[] { (byte) 0xEF, (byte) 0xBF, (byte) 0xBD };
+
+  private static void appendReplacementCharacter(ArrayList<Byte> bytes) {
+    for (byte b : UTF8String.UNICODE_REPLACEMENT_CHARACTER) bytes.add(b);
+  }
+
+  /**
+   * Returns a validated version of the current UTF-8 string by replacing invalid UTF-8 sequences
+   * with the Unicode replacement character (U+FFFD), as per the rules defined in the Unicode
+   * standard. This behaviour is consistent with the behaviour of `UnicodeString` in ICU4C.
+   *
+   * @return A new UTF8String that is a valid UTF8 byte sequence.
+   */
+  public UTF8String makeValidUTF8() {
+    ArrayList<Byte> bytes = new ArrayList<>();
+    int byteIndex = 0;
+    byteIteration:
+    while (byteIndex < numBytes) {
+      // Read the first byte.
+      byte firstByte = getByte(byteIndex);
+      int expectedLen = bytesOfCodePointInUTF8[firstByte & 0xFF];
+      int codePointLen = Math.min(expectedLen, numBytes - byteIndex);
+      // 0B UTF-8 sequence (invalid first byte).
+      if (codePointLen == 0) {
+        appendReplacementCharacter(bytes);
+        byteIndex += 1;
+        continue;
+      }
+      // 1B UTF-8 sequence (ASCII or invalid).
+      if (codePointLen == 1) {
+        if (firstByte >= 0) bytes.add(firstByte);
+        else appendReplacementCharacter(bytes);
+        byteIndex += 1;
+        continue;
+      }
+      // Read the second byte.
+      byte secondByte = getByte(byteIndex + 1);
+      if (!isValidSecondByte(secondByte, firstByte)) {
+        appendReplacementCharacter(bytes);
+        byteIndex += 1;
+        continue;
+      }
+      // Read remaining continuation bytes.
+      int continuationBytes = 2;
+      for (; continuationBytes < codePointLen; ++continuationBytes) {
+        byte nextByte = getByte(byteIndex + continuationBytes);
+        if (!isValidContinuationByte(nextByte)) {
+          break;
+        }
+      }
+      // Invalid UTF-8 sequence (not enough continuation bytes).
+      if (continuationBytes < expectedLen) {
+        appendReplacementCharacter(bytes);
+        byteIndex += continuationBytes;
+        continue;
+      }
+      // Valid UTF-8 sequence.
+      for (int i = 0; i < codePointLen; ++i) {
+        bytes.add(getByte(byteIndex + i));
+      }
+      byteIndex += codePointLen;
+    }
+    return UTF8String.fromBytes(bytes);
+  }
+
+  /**
+   * Checks if the current UTF8String is valid.
+   *
+   * @return If string represents a valid UTF8 byte sequence.
+   */
+  public boolean isValidUTF8() {
+    return makeValidUTF8().equals(this);
+  }
```

Comment:
This is not as efficient as it could be because we are constructing the valid version of the string and then we compare. It should be very easy to implement this directly:
```java
  public UTF8String isValid() {
    int byteIndex = 0;
    while (byteIndex < numBytes) {
      // Read the first byte.
      byte firstByte = getByte(byteIndex);
      int expectedLen = bytesOfCodePointInUTF8[firstByte & 0xFF];
      int codePointLen = Math.min(expectedLen, numBytes - byteIndex);
      // 0B UTF-8 sequence (invalid first byte).
      if (codePointLen == 0) return false;
      if (codePointLen == 1) {
        ++byteIndex;
        continue;
      }
      // Read the second byte.
      byte secondByte = getByte(byteIndex + 1);
      if (!isValidSecondByte(secondByte, firstByte)) return false;
      // Read remaining continuation bytes.
      int continuationBytes = 2;
      for (; continuationBytes < codePointLen; ++continuationBytes) {
        byte nextByte = getByte(byteIndex + continuationBytes);
        if (!isValidContinuationByte(nextByte)) return false;
      }
      // Invalid UTF-8 sequence (not enough continuation bytes).
      if (continuationBytes < expectedLen) return false;
      byteIndex += codePointLen;
    }
    return true;
  }
```

Reply:
there was a lot of copy-paste, and the worst case complexity is the same

but I added this for isValid() now

Reply:
> there was a lot of copy-paste, and the worst case complexity is the same

I agree.

> but I added this for isValid() now

Thank you!